### DISCUSSION TITLE
Recommendation dev environment

### DISCUSSION THREAD

**Discussion by mx03** [advplyr/audiobookshelf#259] [General]
How is the best development environment for this project.

I use wsl2 and restart the backend on every change, but im not sure how work with the frontend without rebuild it on every change.

**Comment by advplyr** [comment]
This still needs to be added to the readme. @tvories wrote a great dev setup guide for the mobile app https://github.com/advplyr/audiobookshelf-app#contributing, so we can use the same template for the server/web app guide.

When I'm working on the server & web app I run them separately in 2 different tabs so that I can get the benefits of Hot Module Replacement in the client.
I'm putting the full setup info here so with some refinements it can eventually go in the readme.

### Requirements
1. Node v24
2. FFmpeg (https://www.ffmpeg.org/download.html)
3. Tone (https://github.com/sandreas/tone)

### Setup
1. Git clone & cd into project dir
2. Create a new file called `dev.js` in the project root dir and copy the following:

```js
// Using port 3333 is important when running the client web app separately
const Path = require('path')
module.exports.config = {
  Port: 3333,
  ConfigPath: Path.resolve('config'),
  MetadataPath: Path.resolve('metadata'),
  FFmpegPath: 'C:\\FFmpeg\\bin\\ffmpeg.exe',
  FFProbePath: 'C:\\FFmpeg\\bin\\ffprobe.exe'
}
```

3. Install server dependencies: `npm ci`
4. cd into the client directory `cd client`
5. Install dependencies for the web app: `npm ci`
6. Generate web app static files: `npm run generate`
7. `cd ..` back to the project root dir
8. start server: `npm run dev`
9. running locally now at `localhost:3333`

(optional) If you want the benefit of auto reload, which is a _must_ for me.
1. Open a new terminal
2. cd into the client directory `.../audiobookshelf/client`
3. start client: `npm run dev`
4. running locally now at default port 3000: `localhost:3000`

When doing it that way you don't need to worry about `localhost:3333`, the web app is already setup to use that as the server. Which means when running the client locally you need to have the server running on port 3333.

**Reply by jflattery** [reply]
should be updated to include using node v16 :D

**Reply by asheingold** [reply]
This should be built using Node 16, not Node 16 or higher.

**Reply by advplyr** [reply]
Thanks updated

**Reply by advplyr** [reply]
The project has just been updated to use node v20

**Reply by advplyr** [reply]
The project has just been updated to use node v24

**Comment by mx03** [comment]
The path you use is `FFmpegPath: 'C:\\FFmpeg\\bin\\ffmpeg.exe'` does audiobookshelf works directly on windows?

**Reply by advplyr** [reply]
Yes, I develop in windows primarily and run my servers in linux.

I am still looking into the best way to package the software for running on a windows server.

**Reply by tvories** [reply]
> I am still looking into the best way to package the software for running on a windows server.

I really like my workflow of using WSL in Windows with Docker/Podman.  It works really well.

**Reply by mx03** [reply]
For creating a package for windows server just use a setup creator like inno setup and [winsw](https://github.com/winsw/winsw) for creating a service.
For updates and install the windows package manager could used [winget](https://github.com/microsoft/winget-cli) .

**Reply by advplyr** [reply]
Yeah I think most Windows servers will still run the docker container at this point. For those not wanting to run docker or cannot run docker because they are using older windows server then we need an executable for them.

I can package the server & web app into an executable, but it still needs to run as a service and have a way to configure the environment variables. Winget isnt supported before windows 10, which could be problematic. I can also see how to wrap the executable with winsw and run it as a service, but I don't see how this could then be packaged into another executable.
I have a few ideas on this, but none of them are pretty. I'm open to any ideas.

If you run `npm run build-win` I setup a command to package the windows exe, but you may need to install some more dependencies.

**Reply by AnthonyBe** [reply]
I came here hoping to find a way to run this natively on a Windows machine (Win 10 or Server 2012 R2 in my case) under Node.js or similar. It sounds like this is still being discussed as to the way you might want to achieve this?

**Reply by advplyr** [reply]
I run the project locally on Win 10 because that is my development environment. You can follow the guide in the post above: https://github.com/advplyr/audiobookshelf/discussions/259#discussioncomment-1869729. You can skip the optional part since that is just for ease of development.

**Reply by advplyr** [reply]
I realize why you mentioned that now. I was just referring to a way for Win server users to install the app without having Node.js and individually installing FFmpeg. If you don't mind doing that setup and running node apps then it'll work fine.

**Reply by AnthonyBe** [reply]
Thanks for the link to the other comment - I hadn't seen that. I'll give it a try!

**Comment by izyspania** [comment]
Hi , it seems that there is a bug when im trying to scan my audiobooks.

[2022-02-02T13:56:06.898Z]  INFO: [Server] Starting Scan
[2022-02-02T13:56:06.899Z]  INFO: [Scanner] Starting library scan lscan_0pr2hzpsk80y6gk0gh for Main
[2022-02-02T13:56:06.909Z] ERROR: [AudioFileScanner] Error: spawn ffprobe ENOENT : "E:/audiobookshelf/audiobooks/Alexandrescu Sorin/Paradoxul - Sorin Alexandrescu/Audio track 1.mp3"
[2022-02-02T13:56:06.910Z] ERROR: [AudioFileScanner] Error: spawn ffprobe ENOENT : "E:/audiobookshelf/audiobooks/Alexandrescu Sorin/Paradoxul - Sorin Alexandrescu/Audio track 2.mp3"
[2022-02-02T13:56:06.910Z]  WARN: [Scanner] Audiobook has no ebooks and no valid audio tracks "Alexandrescu Sorin/Paradoxul - Sorin Alexandrescu"
[2022-02-02T13:56:06.911Z] DEBUG: [DB] Inserted 0 audiobook
[2022-02-02T13:56:06.911Z]  INFO: [Scanner] Library scan lscan_0pr2hzpsk80y6gk0gh completed in 0:00.0 | 0 Added | 0 Updated | 0 Missing
[2022-02-02T13:56:06.911Z]  INFO: [Server] Scan complete
E:\audiobookshelf\node_modules\njodb\index.js:103
        throw error;
        ^

SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at ChildProcess.<anonymous> (E:\audiobookshelf\node_modules\node-ffprobe\lib\ffprobe.js:36:40)
    at ChildProcess.emit (node:events:390:28)
    at maybeClose (node:internal/child_process:1064:16)
    at Socket.<anonymous> (node:internal/child_process:450:11)
    at Socket.emit (node:events:390:28)
    at Pipe.<anonymous> (node:net:687:12)

Edit: i think it doesnt find my FFmpeg

**Reply by advplyr** [reply]
Please try with other audio files

**Reply by izyspania** [reply]
Ok , it was the FFmpeg , i reinstalled it  (setx /m PATH "C:\ffmpeg\bin;%PATH%"), opened a new cmd and now it seems to work. ( i rushed with question before looking into it closer but maybe it will help someone else)

**Comment by jflattery** [comment]
I wanted to share the command I have been using to build:
```bash
ABS_Build_Dir=~/jflattery/audiobookshelf/
cd ${ABS_Build_Dir} &&\
 rm -rf ${ABS_Build_Dir}/node_modules/ &&\
 rm -rf ${ABS_Build_Dir}/client/node_modules/ &&\
 npm ci && npm update && npm dedupe && \
 cd ${ABS_Build_Dir}/client &&\
 npm ci && npm update && npm dedupe && \
 npm run generate &&\
 cd $ABS_Build_Dir &&\
 npm run dev 
```

**Reply by advplyr** [reply]
One thing to keep in mind with the `npm update` is it will update all the dependencies to their latest versions (that match the semver in `package.json`) instead of using the exact versions specified in `package-lock.json`. Just throwing that in since it could throw some people off at first.

**Comment by skanga** [comment]
I see you have a .exe file on Windows in the packages. How do you create this .exe file on Windows?

**Reply by advplyr** [reply]
In the package.json file there is a script `build-win` that uses [pkg](https://www.npmjs.com/package/pkg). Pkg is also used to build the debian file.
If you have the project locally you can run `npm run build-win`

**Comment by TheSameCat2** [comment]
I've been getting this error following the above directions, upon the step 6: ```npm run generate```
```sh: line 1: /media/blockstorage/audiobookshelf-2.2.14/client/node_modules/.bin/nuxt: Permission denied```

Using Ubuntu 22.04, Node 16.19.0
I've used Google. Tried both the tarball of the last release, and main branch.
Any help is deeply appreciated and thanks in advance.

**Reply by TheSameCat2** [reply]
I should mention in advance, I'm only running the project this way because I'm hosting on Arm64v8 on Oracle Cloud. Otherwise I would have used the Docker container.

**Reply by advplyr** [reply]
I don't know what that error is about

**Comment by twdzx** [comment]
In the current version (2.2.18), importing file prompts in txt or TXT format is not supported.
![bug](https://github.com/advplyr/audiobookshelf/assets/103708945/5ffcc3a3-ca81-477d-86f9-b3db04baae32)

**Reply by twdzx** [reply]
I just upgraded 2.2.20, and the txt file upload continues to report errors.

**Reply by advplyr** [reply]
You can't upload a single text file. That is not a valid audio file or ebook format

**Comment by orybrad** [comment]
How do I obtain the test flight code/key for the app?

**Reply by jflattery** [reply]
"_ Beta is currently full. Apple has a hard limit of 10k beta testers. Updates will be posted in Discord/Matrix._"

**Comment by HamKDaBu11** [comment]
Hey everybody! I have setup the dev container and containerized the application on Docker. After that, I am running the command `npm run dev` but I am getting an error that says that ffmpeg and ffprobe versions are too old.....anyone have a solution to this?

**Reply by advplyr** [reply]
Hey if you pull the most recent commit I just updated it so the devcontainer skips the ffmpeg binaries check. Let me know if that fixes it for you.
